### PR TITLE
-bytespersigop option to additionally limit sigops in transactions we relay and mine

### PR DESCRIPTION
**Pull request by luke-jr**

### CODE REVIEW

**Comment by NicolasDorier** [comment]
glad to see that at least we would have a mitigation path in care of sigop spam attack. The worried me a bit.

**Comment by NicolasDorier** [comment]
What is the rational on the default of 20 though ?

During current rules, if all transactions max out the MAX_STANDARD_TX_SIGOPS we can get 5 transactions in the block.

In size, the maximum those 5 transactions would be is 20000 bytes. 
This is 5 bytesperop currently. (without this option)

Not sure if it is relevant though. Just interested to know why you chose 20.

**Comment by jonasschnelli** [comment]
Concept ACK

**Comment by jgarzik** [comment]
ut ACK

**Comment by laanwj** [comment]
Concept ACK

File: src/init.cpp [hunk 1]
```diff
@@ -449,6 +449,7 @@ std::string HelpMessage(HelpMessageMode mode)
     strUsage += HelpMessageGroup(_("Node relay options:"));
     if (showDebug)
         strUsage += HelpMessageOpt("-acceptnonstdtxn", strprintf("Relay and mine \"non-standard\" transactions (%sdefault: %u)", "testnet/regtest only; ", !Params(CBaseChainParams::TESTNET).RequireStandard()));
+    strUsage += HelpMessageOpt("-bytespersigop", strprintf(_("Maximum rate of sigops per byte in transactions we relay and mine (default: %u)"), nBytesPerSigOp));
```
Comment (sipa) [inline src/init.cpp]: Can you use a constant DEFAULT_MAX_BYTES_PER_SIGOP instead here?

**Comment by sipa** [comment]
utACK

**Comment by sipa** [comment]
Needs trivial rebase.

**Comment by luke-jr** [comment]
Nit addressed, and rebased.

**Comment by osogama** [comment]
Ggggggbgh

On Saturday, November 28, 2015, Luke-Jr <notifications@github.com> wrote:

> Nit addressed, and rebased.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/bitcoin/bitcoin/pull/7081#issuecomment-160347898>.
>


-- 
Sent from Gmail Iphone

**Comment by sdaftuar** [comment]
I haven't thought hard enough yet about the value of 20 as a default, but conceptually I think something like this makes sense.  Could you add a unit test or RPC test that exercises the new logic?

**Comment by dcousens** [comment]
utACK

**Comment by gmaxwell** [comment]
ACK, with nit. Do something about that help message.

File: src/init.cpp [hunk 2]
```diff
@@ -475,6 +475,7 @@ std::string HelpMessage(HelpMessageMode mode)
     strUsage += HelpMessageGroup(_("Node relay options:"));
     if (showDebug)
         strUsage += HelpMessageOpt("-acceptnonstdtxn", strprintf("Relay and mine \"non-standard\" transactions (%sdefault: %u)", "testnet/regtest only; ", !Params(CBaseChainParams::TESTNET).RequireStandard()));
+    strUsage += HelpMessageOpt("-bytespersigop", strprintf(_("Maximum rate of sigops per byte in transactions we relay and mine (default: %u)"), DEFAULT_BYTES_PER_SIGOP));
```
Comment (MarcoFalke) [inline src/init.cpp]: Nit: per @gmaxwell 

@luke-jr  Also needs rebase

**Comment by luke-jr** [comment]
Done

**Comment by morcos** [comment]
utACK but agree with previous nit that it would be nice to have a brief explanation of the choice of default.

**Comment by MarcoFalke** [comment]
@gmaxwell Looks like you assigned the wrong milestone.

**Comment by dcousens** [comment]
re-ACK

**Comment by laanwj** [comment]
Changed milestone back to 0.12, added "Needs backport" label as it needs backport to 0.11

File: src/main.cpp [hunk 3]
```diff
@@ -967,6 +960,15 @@ bool AcceptToMemoryPool(CTxMemPool& pool, CValidationState &state, const CTransa
...
 
+        // Check that the transaction doesn't have an excessive number of
+        // sigops, making it impossible to mine. Since the coinbase transaction
+        // itself can contain sigops MAX_STANDARD_TX_SIGOPS is less than
+        // MAX_BLOCK_SIGOPS; we still consider this an invalid rather than
+        // merely non-standard transaction.
+        if ((nSigOps > MAX_STANDARD_TX_SIGOPS) || (nBytesPerSigOp && nSigOps > nSize / nBytesPerSigOp))
+            return state.DoS(0, false, REJECT_NONSTANDARD, "bad-txns-too-many-sigops", false,
+                strprintf("%d", nSigOps));
+
```
Comment (MarcoFalke) [inline src/main.cpp]: @luke-jr Needs rebase

Comment (MarcoFalke) [inline src/main.cpp]: (or merge commit to solve the conlfict)

**Comment by btcdrak** [comment]
utACK 45b8e27 but would be nice to see some RPC tests.

**Comment by laanwj** [comment]
@luke-jr Not sure what you want to do here instead of rebasing, but this somehow needs to be made suitable for merge if you still want it included in 0.12

**Comment by laanwj** [comment]
This has been backported to 0.12 via #7323, removing "needs backport" label

**Comment by ScroogeMcDuckButWithBitcoin** [comment]
So, are we metacoin guys supposed to go back to p2pkh encoding - or should we just fragment our user's wallets into more inputs, before continuing to use op_multisig encodings?

**Comment by morcos** [comment]
There were repeated requests in this PR for a justification of the limit chosen which were ignored.  It was a mistake to merge the PR without first understanding why 20 was the right number.  I apologize for not being more forceful about it at the time.

**Comment by sipa** [comment]
I think we should change the behavior so that transactions with too high
sigops are simply treated as if their corresponding size was larger too.

**Comment by ScroogeMcDuckButWithBitcoin** [comment]
For what it's worth, I've reverted to pub key hash encoding for the time being: https://github.com/17Q4MX2hmktmpuUKHFuoRmS5MfB5XPbhod/dropzone_ruby/blob/master/lib/dropzone/connection.rb#L167-L173

I'd rather not engage in a cat and mouse game on this, and I would expect that doing so would probably just further waste the resources that you generally try to protect.

**Pull request merged**